File: ext/websocket_driver/com/jcoglan/websocket/Parser.java
package com.jcoglan.websocket;

import java.util.Arrays;

public class Parser {
    private static final int
        FIN                   = 0x80,
        RSV1                  = 0x40,
        RSV2                  = 0x20,
        RSV3                  = 0x10,
        OPCODE                = 0x0f,
        MASK                  = 0x80,
        LENGTH                = 0x7f,

        OPCODE_CONTINUTATION  = 0,
        OPCODE_TEXT           = 1,
        OPCODE_BINARY         = 2,
        OPCODE_CLOSE          = 8,
        OPCODE_PING           = 9,
        OPCODE_PONG           = 10,

        NORMAL_CLOSURE        = 1000,
        GOING_AWAY            = 1001,
        PROTOCOL_ERROR        = 1002,
        UNACCEPTABLE          = 1003,
        ENCODING_ERROR        = 1007,
        POLICY_VIOLATION      = 1008,
        TOO_LARGE             = 1009,
        EXTENSION_ERROR       = 1010,
        UNEXPECTED_CONDITION  = 1011,

        DEFAULT_ERROR_CODE    = 1000,
        MIN_RESERVED_ERROR    = 3000,
        MAX_RESERVED_ERROR    = 4999;

    private boolean requireMasking;

    private StreamReader reader;
    private Extensions extensions;
    private Observer observer;

    private int stage;
    private Frame frame;
    private Message message;

    private int errorCode;
    private String errorReason;

    public Parser(Extensions extensions, Observer observer, boolean requireMasking) {
        reader = new StreamReader();

        this.requireMasking = requireMasking;
        this.extensions = extensions;
        this.observer = observer;

        stage = 1;
        frame = null;
        message = null;

        errorCode = 0;
        errorReason = null;
    }

    public void parse(byte[] data) {
        reader.push(data);
        byte[] chunk = new byte[0];

        while (chunk != null) {
            switch (stage) {
                case 1:
                    chunk = reader.read(2);
                    if (chunk != null) parseHead(chunk);
                    break;
                case 2:
                    chunk = reader.read(frame.lengthBytes);
                    if (chunk != null) parseExtendedLength(chunk);
                    break;
                case 3:
                    chunk = reader.read(4);
                    if (chunk != null) {
                        stage = 4;
                        frame.maskingKey = chunk;
                    }
                    break;
                case 4:
                    chunk = reader.read((int)frame.length);
                    if (chunk != null) {
                        stage = 1;
                        emitFrame(chunk);
                    }
                    break;
                default:
                    chunk = null;
                    break;
            }
        }
    }

    private void parseError(int code, String reason) {
        if (errorCode != 0) return;

        stage       = 0;
        errorCode   = code;
        errorReason = reason;

        observer.onError(code, reason);
    }

    private void parseHead(byte[] chunk) {
        frame = new Frame();

        frame.fin    = (chunk[0] & FIN)  == FIN;
        frame.rsv1   = (chunk[0] & RSV1) == RSV1;
        frame.rsv2   = (chunk[0] & RSV2) == RSV2;
        frame.rsv3   = (chunk[0] & RSV3) == RSV3;
        frame.opcode = (chunk[0] & OPCODE);
        frame.masked = (chunk[1] & MASK) == MASK;
        frame.length = (chunk[1] & LENGTH);

        if (!extensions.validFrameRsv(frame.rsv1, frame.rsv2, frame.rsv3, frame.opcode)) {
            parseError(PROTOCOL_ERROR, 
                String.format("One or more reserved bits are on: reserved1 = %d, reserved2 = %d, reserved3 = %d",
                    frame.rsv1 ? 1 : 0,
                    frame.rsv2 ? 1 : 0,
                    frame.rsv3 ? 1 : 0));
            return;
        }

        if (!validOpcode(frame.opcode)) {
            parseError(PROTOCOL_ERROR, String.format("Unrecognized frame opcode: %d", frame.opcode));
            return;
        }

        if (controlOpcode(frame.opcode) && !frame.fin) {
            parseError(PROTOCOL_ERROR, String.format("Received fragmented control frame: opcode = %d", frame.opcode));
            return;
        }

        if (message == null && frame.opcode == OPCODE_CONTINUTATION) {
            parseError(PROTOCOL_ERROR, "Received unexpected continuation frame");
            return;
        }

        if (message != null && openingOpcode(frame.opcode)) {
            parseError(PROTOCOL_ERROR, "Received new data frame but previous continuous frame is unfinished");
            return;
        }

        if (requireMasking && !frame.masked) {
            parseError(UNACCEPTABLE, "Received unmasked frame but masking is required");
            return;
        }

        if (frame.length <= 125) {
            if (!checkFrameLength()) return;
            stage = frame.masked ? 3 : 4;
        } else {
            stage = 2;
            frame.lengthBytes = (frame.length == 126) ? 2 : 8;
        }
    }

    private boolean validOpcode(int opcode) {
        return controlOpcode(opcode) || messageOpcode(opcode);
    }

    private boolean controlOpcode(int opcode) {
        return opcode == OPCODE_CLOSE ||
               opcode == OPCODE_PING ||
               opcode == OPCODE_PONG;
    }

    private boolean messageOpcode(int opcode) {
        return openingOpcode(opcode) || opcode == OPCODE_CONTINUTATION;
    }

    private boolean openingOpcode(int opcode) {
        return opcode == OPCODE_TEXT ||
               opcode == OPCODE_BINARY;
    }

    private void parseExtendedLength(byte[] chunk) {
        if (frame.length == 126) {
            frame.length = Buffer.readUInt16(chunk, 0);
        } else if (frame.length == 127) {
            frame.length = Buffer.readUInt64(chunk, 0);
        }

        if (controlOpcode(frame.opcode) && frame.length > 125) {
            parseError(PROTOCOL_ERROR, String.format("Received control frame having too long payload: %d", frame.length));
            return;
        }

        if (!checkFrameLength()) return;

        stage = frame.masked ? 3 : 4;
    }

    private boolean checkFrameLength() {
        if (Message.wouldOverflow(message, frame)) {
            parseError(TOO_LARGE, "WebSocket frame length too large");
            return false;
        } else {
            return true;
        }
    }

    private void emitFrame(byte[] chunk) {
        frame.payload = chunk;
        Frame.mask(frame);

        stage = 1;

        int code = 0;
        byte[] reason = new byte[0];

        switch (frame.opcode) {
            case OPCODE_CONTINUTATION:
                message.push(frame);
                break;

            case OPCODE_TEXT:
            case OPCODE_BINARY:
                message = new Message(frame);
                break;

            case OPCODE_CLOSE:
                if (frame.length == 0) {
                    code   = DEFAULT_ERROR_CODE;
                    reason = new byte[0];
                } else if (frame.length >= 2) {
                    code   = (int)Buffer.readUInt16(frame.payload, 0);
                    reason = Arrays.copyOfRange(frame.payload, 2, (int)frame.length);
                }

                if (!validCloseCode(code)) {
                    code = PROTOCOL_ERROR;
                }
                observer.onClose(code, reason);
                break;

            case OPCODE_PING:
                observer.onPing(frame);
                break;

            case OPCODE_PONG:
                observer.onPong(frame);
                break;
        }

        if (frame.opcode <= OPCODE_BINARY && frame.fin) emitMessage();
        frame = null;
    }

    private boolean validCloseCode(int code) {
        return code == NORMAL_CLOSURE ||
               code == GOING_AWAY ||
               code == PROTOCOL_ERROR ||
               code == UNACCEPTABLE ||
               code == ENCODING_ERROR ||
               code == POLICY_VIOLATION ||
               code == TOO_LARGE ||
               code == EXTENSION_ERROR ||
               code == UNEXPECTED_CONDITION ||
               (code >= MIN_RESERVED_ERROR && code <= MAX_RESERVED_ERROR);
    }

    private void emitMessage() {
        observer.onMessage(message);
        message = null;
    }
}

File: lib/websocket/driver/hybi/parser.rb
module WebSocket
  class Driver
    class Hybi

      class Parser
        def initialize(driver, require_masking)
          @driver = driver
          @require_masking = require_masking

          @reader  = StreamReader.new
          @stage   = 0
          @frame   = nil
          @message = nil
        end

        def parse(chunk)
          @reader.push(chunk)
          buffer = true

          while buffer
            case @stage
              when 0 then
                buffer = @reader.read(1)
                parse_opcode(buffer.getbyte(0)) if buffer

              when 1 then
                buffer = @reader.read(1)
                parse_length(buffer.getbyte(0)) if buffer

              when 2 then
                buffer = @reader.read(@frame.length_bytes)
                parse_extended_length(buffer) if buffer

              when 3 then
                buffer = @reader.read(4)
                if buffer
                  @stage = 4
                  @frame.masking_key = buffer
                end

              when 4 then
                buffer = @reader.read(@frame.length)

                if buffer
                  @stage = 0
                  emit_frame(buffer)
                end

              else
                buffer = nil
            end
          end
        end

      private

        def parse_opcode(octet)
          rsvs = [RSV1, RSV2, RSV3].map { |rsv| (octet & rsv) == rsv }

          @frame = Frame.new

          @frame.final  = (octet & FIN) == FIN
          @frame.rsv1   = rsvs[0]
          @frame.rsv2   = rsvs[1]
          @frame.rsv3   = rsvs[2]
          @frame.opcode = (octet & OPCODE)

          @stage = 1

          unless @driver.__send__(:valid_frame_rsv?, @frame.rsv1, @frame.rsv2, @frame.rsv3, @frame.opcode)
            return parser_error(:protocol_error,
                "One or more reserved bits are on: reserved1 = #{@frame.rsv1 ? 1 : 0}" +
                ", reserved2 = #{@frame.rsv2 ? 1 : 0 }" +
                ", reserved3 = #{@frame.rsv3 ? 1 : 0 }")
          end

          unless OPCODES.values.include?(@frame.opcode)
            return parser_error(:protocol_error, "Unrecognized frame opcode: #{@frame.opcode}")
          end

          unless MESSAGE_OPCODES.include?(@frame.opcode) or @frame.final
            return parser_error(:protocol_error, "Received fragmented control frame: opcode = #{@frame.opcode}")
          end

          if @message and OPENING_OPCODES.include?(@frame.opcode)
            return parser_error(:protocol_error, 'Received new data frame but previous continuous frame is unfinished')
          end
        end

        def parse_length(octet)
          @frame.masked = (octet & MASK) == MASK
          @frame.length = (octet & LENGTH)

          if @frame.length >= 0 and @frame.length <= 125
            @stage = @frame.masked ? 3 : 4
            return unless check_frame_length
          else
            @stage = 2
            @frame.length_bytes = (@frame.length == 126) ? 2 : 8
          end

          if @require_masking and not @frame.masked
            return parser_error(:unacceptable, 'Received unmasked frame but masking is required')
          end
        end

        def parse_extended_length(buffer)
          @frame.length = buffer.unpack(PACK_FORMATS[buffer.bytesize]).first
          @stage = @frame.masked ? 3 : 4

          unless MESSAGE_OPCODES.include?(@frame.opcode) or @frame.length <= 125
            return parser_error(:protocol_error, "Received control frame having too long payload: #{@frame.length}")
          end

          return unless check_frame_length
        end

        def check_frame_length
          length = @message ? @message.data.bytesize : 0

          if length > MAX_MESSAGE_LENGTH - @frame.length
            parser_error(:too_large, 'WebSocket frame length too large')
            false
          else
            true
          end
        end

        def emit_frame(buffer)
          frame    = @frame
          opcode   = frame.opcode
          payload  = frame.payload = Mask.mask(buffer, @frame.masking_key)
          bytesize = payload.bytesize
          bytes    = payload.bytes.to_a

          @frame = nil

          code   = 0
          reason = nil

          case opcode
            when OPCODES[:continuation] then
              return parser_error(:protocol_error, 'Received unexpected continuation frame') unless @message
              @message << frame

            when OPCODES[:text], OPCODES[:binary] then
              @message = Message.new
              @message << frame

            when OPCODES[:close] then
              if frame.length == 0
                code   = DEFAULT_ERROR_CODE
                reason = ''
              elsif frame.length >= 2
                code   = payload.unpack(PACK_FORMATS[2]).first
                reason = bytes[2..-1]
              end

              unless ERROR_CODES.include?(code) or
                     (code >= MIN_RESERVED_ERROR and code <= MAX_RESERVED_ERROR)
                code = ERRORS[:protocol_error]
              end

              @driver.__send__(:handle_close, code, reason || '')

            when OPCODES[:ping] then
              @driver.__send__(:handle_ping, payload)

            when OPCODES[:pong] then
              @driver.__send__(:handle_pong, payload)
          end

          emit_message if frame.final and MESSAGE_OPCODES.include?(opcode)
        end

        def emit_message
          @driver.__send__(:handle_message, @message.opcode, @message.rsv1, @message.rsv2, @message.rsv3, @message.data)
          @message = nil
        end

        def parser_error(type, message)
          @stage = -1
          @driver.__send__(:handle_error, ERRORS[type], message)
        end
      end

    end
  end
end
